chapter_recurrent-modern/gru.ipynb

```diff
@@ -0,0 +1,482 @@
+    "\n",
+    "## Implementation from Scratch\n",
+    "\n",
+    "To gain a better understanding of the GRU model, let us implement it from scratch. We begin by reading the time machine dataset that we used in :numref:`sec_rnn_scratch`. The code for reading the dataset is given below.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%mavenRepo snapshots https://oss.sonatype.org/content/repositories/snapshots/\n",
+    "%maven org.slf4j:slf4j-api:1.7.26\n",
+    "%maven org.slf4j:slf4j-simple:1.7.26\n",
+    "    \n",
+    "%maven ai.djl:api:0.11.0-SNAPSHOT\n",
+    "%maven ai.djl:basicdataset:0.11.0-SNAPSHOT\n",
+    "\n",
+    "// See https://github.com/awslabs/djl/blob/master/mxnet/mxnet-engine/README.md\n",
+    "// MXNet \n",
+    "%maven ai.djl.mxnet:mxnet-engine:0.11.0-SNAPSHOT\n",
+    "%maven ai.djl.mxnet:mxnet-native-auto:1.7.0-backport"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%load ../utils/plot-utils\n",
+    "%load ../utils/Functions.java\n",
+    "%load ../utils/PlotUtils.java\n",
+    "\n",
+    "%load ../utils/StopWatch.java\n",
+    "%load ../utils/Accumulator.java\n",
+    "%load ../utils/Animator.java\n",
+    "%load ../utils/Training.java\n",
+    "%load ../utils/TimeMachineUtils.java"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "NDManager manager = NDManager.newBaseManager(Functions.tryGpu(0));"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "int batchSize = 32;\n",
+    "int numSteps = 35;\n",
+    "\n",
+    "TimeMachineDataset dataset =\n",
+    "        new TimeMachineDataset.Builder()\n",
+    "                .setManager(manager)\n",
+    "                .setMaxTokens(10000)\n",
+    "                .setSampling(batchSize, false)\n",
+    "                .setSteps(numSteps)\n",
+    "                .build();\n",
+    "dataset.prepare();\n",
+    "Vocab vocab = dataset.getVocab();"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "origin_pos": 3
+   },
+   "source": [
+    "### Initializing Model Parameters\n",
+    "\n",
+    "The next step is to initialize the model parameters.\n",
+    "We draw the weights from a Gaussian distribution\n",
+    "with standard deviation to be 0.01 and set the bias to 0. The hyperparameter `num_hiddens` defines the number of hidden units.\n",
+    "We instantiate all weights and biases relating to the update gate, the reset gate, the candidate hidden state,\n",
+    "and the output layer."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "public static NDArray normal(Shape shape, Device device) {\n",
```

Comment:
actually DJL will detect system GPU and use it if applicable. So I think you can remove the device parameter here. If you check TensorFlow implementation, they don't have this param as it do it in DJL way as well.

Reply:
Ohh I just saw randomNormal uses the NDArray device by default if you don't specify. Cool, I'll remove it.

Reply:
I'll be doing this in another PR as it covers multiple sections